Bug: parserOptions.project with single-run mode and extraFileExtensions: files not in program instance(s)

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

After updating https://github.com/ota-meshi/eslint-plugin-yml to `8.0.0-alpha.40`, trying to run `npm run lint` gives:

```plaintext
/Users/josh/repos/eslint-plugin-yml/docs/.vitepress/theme/components/components/EslintPluginEditor.vue
  0:0  error  Parsing error: "parserOptions.programs" has been provided for @typescript-eslint/parser.
The file was not found in any of the provided program instance(s): docs/.vitepress/theme/components/components/EslintPluginEditor.vue
```

...for six files in total. Linting works fine if you switch to `projectService: true`. All three variations of linting take about 130s on my Mac Studio. I posted a comment here: https://github.com/ota-meshi/eslint-plugin-yml/pull/346/files#r1667218685.

...and for a second, quicker reproduction (~30s), here: https://github.com/ota-meshi/eslint-plugin-jsonc/pull/351/files#r1667375018

### Reproduction Repository Link

* https://github.com/ota-meshi/eslint-plugin-jsonc/pull/351
* https://github.com/ota-meshi/eslint-plugin-yml/pull/346

### Repro Steps

1) clone https://github.com/ota-meshi/eslint-plugin-yml
2) `gh pr checkout 346`
3) `npm install`
4) `npm run lint`


### Versions


<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.0.0-alpha.39` |
| `@typescript-eslint/parser`            | `8.0.0-alpha.39` |
| `@typescript-eslint/scope-manager`     | `8.0.0-alpha.39` |
| `@typescript-eslint/typescript-estree` | `8.0.0-alpha.39` |
| `@typescript-eslint/type-utils`        | `8.0.0-alpha.39` |
| `@typescript-eslint/utils`             | `8.0.0-alpha.39` |
| `TypeScript`                           | `5.5.0` |
| `ESLint`                               | `9.5.0` |

💖

Hmm... When `eslint` is invoked without params, it still somehow fails:

```console
$ TSESTREE_SINGLE_RUN=false eslint

.../src/App.vue
  0:0  error  Parsing error: ESLint was configured to run on `<tsconfigRootDir>/src/App.vue` using `parserOptions.project`: 
- <tsconfigRootDir>/tsconfig.app.json
- <tsconfigRootDir>/tsconfig.node.json
However, none of those TSConfigs include this file. Either:
- Change ESLint's list of included files to not include this file
- Change one of those TSConfigs to include this file
- Create a new TSConfig that includes this file and include it in your parserOptions.project
See the typescript-eslint docs for more info: https://typescript-eslint.io/troubleshooting/typed-linting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file

$ TSESTREE_SINGLE_RUN=false eslint src

.../src/App.vue
  6:1  warning  Unexpected console statement  no-console

✖ 1 problem (0 errors, 1 warning)

$
```

Not a biggie, but still somewhat unexpected 😄 Relevant part of my `eslint.config.mjs`:

```javascript
import tsParser from '@typescript-eslint/parser'
import * as jsParser from 'espree'

export default [
  ///  ... all other configs...
  {
    name: 'vue:parser',
    files: [ '**/*.vue' ],
    languageOptions: {
      parser: vueParser,
      parserOptions: {
        parser: {
          ts: tsParser,
          js: jsParser,
        },
        extraFileExtensions: [ '.vue' ],
      },
    },
  }, {
    languageOptions: {
      parserOptions: {
        createDefaultProgram: false,
        project: [
          './tsconfig.app.json',
          './tsconfig.node.json',
        ],
      },
    },
  },
]
```

--- a/packages/typescript-estree/src/parseSettings/inferSingleRun.ts
+++ b/packages/typescript-estree/src/parseSettings/inferSingleRun.ts
@@ -15,6 +15,13 @@ import type { TSESTreeOptions } from '../parser-options';
  * @returns Whether this is part of a single run, rather than a long-running process.
  */
 export function inferSingleRun(options: TSESTreeOptions | undefined): boolean {
+  // https://github.com/typescript-eslint/typescript-eslint/issues/9504
+  // There's no support (yet?) for extraFileExtensions in single-run hosts.
+  // Only watch program hosts and project service can support that.
+  if (options?.extraFileExtensions?.length) {
+    return false;
+  }
+
   if (
     // single-run implies type-aware linting - no projects means we can't be in single-run mode
     options?.project == null ||

--- a/packages/typescript-estree/tests/lib/inferSingleRun.test.ts
+++ b/packages/typescript-estree/tests/lib/inferSingleRun.test.ts
@@ -63,6 +63,30 @@ describe('inferSingleRun', () => {
     expect(actual).toBe(true);
   });
 
+  it('returns true when singleRun can be inferred and options.extraFileExtensions is an empty array', () => {
+    process.env.CI = 'true';
+
+    const actual = inferSingleRun({
+      allowAutomaticSingleRunInference: true,
+      extraFileExtensions: [],
+      project: './tsconfig.json',
+    });
+
+    expect(actual).toBe(true);
+  });
+
+  it('returns false when singleRun can be inferred options.extraFileExtensions contains entries', () => {
+    process.env.CI = 'true';
+
+    const actual = inferSingleRun({
+      allowAutomaticSingleRunInference: true,
+      extraFileExtensions: ['.vue'],
+      project: './tsconfig.json',
+    });
+
+    expect(actual).toBe(false);
+  });
+
   it('returns false when there is no way to infer singleRun', () => {
     const actual = inferSingleRun({
       programs: null,

